src/script/sign.cpp

```diff
@@ -612,15 +612,18 @@ bool IsSolvable(const SigningProvider& provider, const CScript& script)
 
 bool IsSegWitOutput(const SigningProvider& provider, const CScript& script)
 {
-    std::vector<valtype> solutions;
-    auto whichtype = Solver(script, solutions);
-    if (whichtype == TxoutType::WITNESS_V0_SCRIPTHASH || whichtype == TxoutType::WITNESS_V0_KEYHASH || whichtype == TxoutType::WITNESS_UNKNOWN) return true;
-    if (whichtype == TxoutType::SCRIPTHASH) {
-        auto h160 = uint160(solutions[0]);
-        CScript subscript;
-        if (provider.GetCScript(CScriptID{h160}, subscript)) {
-            whichtype = Solver(subscript, solutions);
-            if (whichtype == TxoutType::WITNESS_V0_SCRIPTHASH || whichtype == TxoutType::WITNESS_V0_KEYHASH || whichtype == TxoutType::WITNESS_UNKNOWN) return true;
+    int version;
+    valtype program;
+    if (script.IsWitnessProgram(version, program)) return true;
+    if (script.IsPayToScriptHash()) {
+        std::vector<valtype> solutions;
+        auto whichtype = Solver(script, solutions);
+        if (whichtype == TxoutType::SCRIPTHASH) {
+            auto h160 = uint160(solutions[0]);
+            CScript subscript;
+            if (provider.GetCScript(CScriptID{h160}, subscript)) {
+                if (subscript.IsWitnessProgram(version, program)) return true;
```

Comment:
This would return true for Taproot scripts that are in the P2SH even though such scripts cannot be spent. Do we want to do that?

Reply:
Yes, I noticed that, but I think my answer is yes. It also used to return true for unknown witness versions, so no reason to exclude v1 p2sh. It's still segwit; just not taproot.